src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetCoreMvc2Integration.cs

```diff
@@ -0,0 +1,311 @@
+            }
+            catch (Exception ex)
+            {
+                // profiled app will continue working as expected without this method
+                Log.Error($"Exception when calling {methodDef}.", ex);
+            }
+            finally
+            {
+                aspNetCoreMvcActionScope?.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Wrapper method used to catch unhandled exceptions in the incoming request pipeline for Microsoft.AspNetCore.Mvc.Core
+        /// </summary>
+        /// <param name="context">The DiagnosticSource that this extension method was called on.</param>
+        /// <param name="opCode">The OpCode used in the original method call.</param>
+        /// <param name="mdToken">The mdToken of the original method call.</param>
+        [InterceptMethod(
+            CallerAssembly = AspnetMvcCore,
+            TargetAssembly = AspnetMvcCore,
+            TargetType = ResourceInvoker,
+            TargetSignatureTypes = new[] { ClrNames.Void, ClrNames.Ignore },
+            TargetMinimumVersion = Major2,
+            TargetMaximumVersion = Major2)]
+        public static void Rethrow(object context, int opCode, int mdToken)
+        {
+            string methodDef = $"{ResourceInvoker}.{nameof(Rethrow)}({context?.GetType().FullName} context)";
+            var shouldTrace = Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName);
+            MethodBase instrumentedMethod;
+
+            try
+            {
+                instrumentedMethod = Assembly.GetCallingAssembly().ManifestModule.ResolveMethod(mdToken);
+            }
+            catch (Exception ex)
+            {
+                // profiled app will not continue working as expected without this method
+                Log.ErrorException($"Error retrieving {methodDef}", ex);
+                throw;
+            }
+
+            AspNetCoreIntegrationContext integration = null;
+            var exceptionToGrab = context.GetProperty<Exception>("Exception");
+
+            if (shouldTrace)
+            {
+                var httpContextResult = context.GetProperty("HttpContext");
+
+                if (httpContextResult.HasValue)
+                {
+                    integration = AspNetCoreIntegrationContext.RetrieveFromHttpContext(httpContextResult.Value);
+                }
+
+                if (integration == null)
+                {
+                    Log.Error($"Could not access {nameof(AspNetCoreIntegrationContext)} for {methodDef}.");
+                }
+            }
+
+            try
+            {
+                // call the original method, catching and rethrowing any unhandled exceptions
+                instrumentedMethod.Invoke(null, new[] { context });
+            }
+            catch (Exception ex) when (integration?.SetExceptionOnRootSpan(exceptionToGrab.HasValue ? exceptionToGrab.Value : ex) ?? false)
```

Comment:
I opt to retrieve the original exception from the mvc pipeline if I'm able to, as it's not polluted by our code.